pkg/kapis/devops/v1alpha3/handler.go

```diff
@@ -248,95 +260,89 @@ func (h *devopsHandler) UpdatePipeline(request *restful.Request, response *restf
-
-	response.WriteEntity(obj)
 }
 
 func (h *devopsHandler) DeletePipeline(request *restful.Request, response *restful.Response) {
 	devops := request.PathParameter("devops")
 	pipeline := request.PathParameter("pipeline")
 
 	klog.V(8).Infof("ready to delete pipeline %s/%s", devops, pipeline)
-	err := h.devops.DeletePipelineObj(devops, pipeline)
 
-	if err != nil {
-		klog.Error(err)
-		if errors.IsNotFound(err) {
-			api.HandleNotFound(response, request, err)
+	if client, err := h.getDevOps(request); err == nil {
+		err := client.DeletePipelineObj(devops, pipeline)
+		if err != nil {
+			klog.Error(err)
+			if errors.IsNotFound(err) {
+				api.HandleNotFound(response, request, err)
+				return
+			}
+			api.HandleBadRequest(response, request, err)
 			return
 		}
+		_ = response.WriteEntity(servererr.None)
+	} else {
 		api.HandleBadRequest(response, request, err)
-		return
 	}
-
-	response.WriteEntity(servererr.None)
 }
 
 //credential handler about get/list/post/put/delete
 func (h *devopsHandler) GetCredential(request *restful.Request, response *restful.Response) {
 	devops := request.PathParameter("devops")
 	credential := request.PathParameter("credential")
 
-	obj, err := h.devops.GetCredentialObj(devops, credential)
+	if client, err := h.getDevOps(request); err == nil {
+		obj, err := client.GetCredentialObj(devops, credential)
 
-	if err != nil {
-		klog.Error(err)
-		if errors.IsNotFound(err) {
-			api.HandleNotFound(response, request, err)
+		if err != nil {
+			klog.Error(err)
+			if errors.IsNotFound(err) {
+				api.HandleNotFound(response, request, err)
+				return
+			}
+			api.HandleBadRequest(response, request, err)
 			return
 		}
-		api.HandleBadRequest(response, request, err)
-		return
-	}
 
-	response.WriteEntity(obj)
-}
-
-func (h *devopsHandler) getDevOps(request *restful.Request) (devops.DevopsOperator, error) {
-	ctx := request.Request.Context()
-	token := ctx.Value(constants.K8SToken).(string)
-
-	if kubernetesClient, err := k8s.NewKubernetesClientWithToken(token, h.k8sClient.Config().Host); err != nil {
-		return nil, err
+		_ = response.WriteEntity(obj)
 	} else {
-		informerFactory := devopsinformers.NewInformerFactories(kubernetesClient.Kubernetes(),
-			kubernetesClient.KubeSphere(),
-			kubernetesClient.ApiExtensions())
-
-		return devops.NewDevopsOperator(h.devopsClient, kubernetesClient.Kubernetes(),
-			kubernetesClient.KubeSphere(),
-			informerFactory.KubeSphereSharedInformerFactory(),
-			informerFactory.KubernetesSharedInformerFactory()), nil
+		api.HandleBadRequest(response, request, err)
 	}
 }
 
 func (h *devopsHandler) ListCredential(request *restful.Request, response *restful.Response) {
 	devops := request.PathParameter("devops")
 	query := query.ParseQueryParameter(request)
-	objs, err := h.devops.ListCredentialObj(devops, query)
 
-	if err != nil {
-		klog.Error(err)
-		if errors.IsNotFound(err) {
-			api.HandleNotFound(response, request, err)
+	if client, err := h.getDevOps(request); err == nil {
```

Comment:
It does not work well. See the following screenshot.

![image](https://user-images.githubusercontent.com/1450685/125225908-a1dd3280-e302-11eb-806c-25f995b4d017.png)

![image](https://user-images.githubusercontent.com/1450685/125225902-9ee24200-e302-11eb-9cdc-356d3f4cebd9.png)

![image](https://user-images.githubusercontent.com/1450685/125225894-98ec6100-e302-11eb-9571-bc1cb2ecfc15.png)

Reply:
I'm sorry for that I don't test most apis manually but pipeline list api. And I also discarded pr https://github.com/LinuxSuRen/ks-devops/pull/29 by mistake.

Reply:
Thanks for your patient review and test.

Reply:
> I'm sorry for that I don't test most apis manually but pipeline list API

I did that work from https://github.com/LinuxSuRen/ks-devops/pull/21/

Please make sure that all changes was tested before others start to review your changes.